python/tank/bootstrap/configuration_writer.py

```diff
@@ -28,6 +28,50 @@
 log = LogManager.get_logger(__name__)
 
 
+class RenameGuard(object):
+    """
+    Ensures that file that were renamed during a scope are renamed to their
+    original name if an exception was raised during that scope.
+    """
+
+    def __init__(self, undo_on_error):
+        """
+        :param bool undo_on_error: If true, the renames will be undone when an exception
+            is raised. If false, the files won't be renamed when an exception is raised.
+        """
+        self._undo_on_error = undo_on_error
+        self._renames = []
+
+    def __enter__(self):
+        """
+        Returns itself so files can be renamed and tracked.
+        """
+        return self
+
+    def rename(self, source, dest):
+        """
+        Renames a file and keeps track of the rename if it succeeded.
+
+        :param str source: File to rename.
+        :param str dest: New name of the file.
+        """
+        os.rename(source, dest)
+        self._renames.append((source, dest))
```

Comment:
So there are three places where we need to handle an exception being thrown.
1. before os.rename succeeds
2. Between the two calls
3. After the second one.

If there is an error before the first, nothing has been moved and nothing needs to be undone.

If there is an error before the rename or during it (rename is atomic, it moves a folder from one location to another), nothing got moved and nothing was added to the _moves list.

If something happened between the two calls (or the second os.rename fails, again), then then only the first path will have been added the only that file will be moved back.